scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -286,14 +286,14 @@
                                         :missing nil
                                         ; The writeback to datomic has occurred, so there's nothing to do except to delete the pod from kubernetes
                                         ; and remove it from our tracking.
-                                        :pod/failed (delete-pod api-client pod)
+                                        :pod/failed (delete-pod api-client cook-expected-state-dict pod)
                                         ; Who resurrected this pod? Where did it come from? Do we have two instances of cook?
                                         :pod/running (kill-pod-in-weird-state compute-cluster pod-name
                                                                               cook-expected-state-dict
                                                                               k8s-actual-state-dict)
                                         ; The writeback to datomic has occurred, so there's nothing to do except to delete the pod from kubernetes
                                         ; and remove it from our tracking.
-                                        :pod/succeeded (delete-pod api-client pod)
+                                        :pod/succeeded (delete-pod api-client cook-expected-state-dict pod)
```

Comment:
This is the bug fix. The old bad code would go:
(completed,succeeded) -> (missing,succeeded) -> (missing,missing), inadvertently deleting its local state first and keeping the kubernetes state.
The correct transitions should go:
(completed, succeeded) -> (completed,missing), (missing,missing).
so we're always tracking the local state.

Reply:
Thanks, this is really helpful.

I think this is context that is super useful to put in the PR description.

Reply:
@scrosby:

Seems like we should add a test to make sure we're now handling this case correctly. 

Specifically:

- assert that we only attempt to delete the pod once (without this, we're attempting 2 deletes)
- assert that we cycle through (completed, missing) when the pod deleted watch message comes in
- assert that we don't cycle through any other state pairs than the correct ones above

Reply:
I can do the first one, I can somewhat do the second, but can't do the last. We recur at the end of processing so some of the intermediate states  aren't visible outside the function. (They can still be visited if we have a failure.)